File: A2/self_tests-JDBC/jdbc/jdbc_tester.py
#!/usr/bin/env python3

import sys
import os
import subprocess
from sql_tester import SQLTester


class JDBCTester(SQLTester):

    def __init__(self, oracle_database, test_database, user_name, user_password, data_files, schema_name,
                 path_to_solution, java_files, java_classpath, need_oracle, output_filename='result_jdbc.txt'):
        super(JDBCTester, self).__init__(oracle_database, test_database, user_name, user_password, data_files,
                                         schema_name, path_to_solution, output_filename)
        self.java_files = java_files
        self.java_classpath = java_classpath
        self.need_oracle = need_oracle

    def init_java(self):
        shell_command = ['javac', '-cp', self.java_classpath, '{}.java'.format(self.__class__.__name__)]
        subprocess.run(shell_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True,
                               check=True)

    def run_java(self, data_file, test_name, output_open):
        shell_command = ['java', '-cp', self.java_classpath, self.__class__.__name__, self.test_database,
                         self.user_name, self.user_password, self.schema_name, data_file, test_name]
        java = subprocess.run(shell_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True,
                              check=True)
        print(java.stdout)
        output_open.write(java.stderr)

    def run(self):

        try:
            for java_file in self.java_files:
                if not os.path.isfile(java_file):
                    self.print_result(name='All JDBC tests', input='', expected='',
                                      actual='File {} not found'.format(java_file), marks=0, status='fail')
                    return
            self.init_java()
            self.init_db()
            with open(self.output_filename, 'w') as output_open:
                for test_name in self.data_files.keys():
                    for data_file in self.data_files[test_name]:
                        data_name = data_file.partition('.')[0]
                        test_data_name = '{} + {}'.format(test_name, data_name)
                        try:
                            # drop + recreate test schema + dataset + fetch test results
                            self.set_test_schema(data_file=data_file)
                            self.run_java(data_file=data_file, test_name=test_name, output_open=output_open)
                            if not self.need_oracle[test_name]:
                                continue
                            sql_file = os.path.join(self.path_to_solution, self.QUERY_DIR,
                                                    '{}_select.sql'.format(test_name))
                            test_results = self.get_test_results(sql_file=sql_file)
                            # fetch results from oracle
                            oracle_results = self.get_oracle_results(data_name=data_name, test_name=test_name)
                            # compare test results with oracle
                            result = self.check_results(oracle_results=oracle_results, test_results=test_results)
                            self.print_result(name=test_data_name, input='', expected='',
                                              actual=result[0] if result[0] else 'All good', marks=result[1],
                                              status=result[2])
                            self.print_result_file(output_open=output_open, test_name=test_data_name,
                                                   message=(result[0] if result[0] else 'All good'),
                                                   oracle_results=oracle_results, test_results=test_results)
                        except subprocess.CalledProcessError as e:
                            self.oracle_connection.commit()
                            self.test_connection.commit()
                            msg = 'Java error\nstdout: {stdout}\nstderr: {stderr}'.format(stdout=e.stdout,
                                                                                          stderr=e.stderr)
                            self.print_result(name=test_data_name, input='', expected='', actual=msg, marks=0,
                                              status='error')
                        except Exception as e:
                            self.oracle_connection.commit()
                            self.test_connection.commit()
                            self.print_result(name=test_data_name, input='', expected='', actual=str(e), marks=0,
                                              status='error')
        except subprocess.CalledProcessError as e:
            msg = 'Java compilation error\nstdout: {stdout}\nstderr: {stderr}'.format(stdout=e.stdout, stderr=e.stderr)
            self.print_result(name='All JDBC tests', input='', expected='', actual=msg, marks=0, status='error')
        except Exception as e:
            self.print_result(name='All JDBC tests', input='', expected='', actual=str(e), marks=0, status='error')
        finally:
            self.close_db()

if __name__ == '__main__':

    PATH_TO_SOLUTION = 'solution'
    DATA_FILES = {'homeownerRecommendation': ['datahr.sql'], 'booking': ['datab.sql']}
    SCHEMA_NAME = 'bnb'
    DATABASE = sys.argv[1]
    USER = sys.argv[2]
    PASSWORD = ''

    JAVA_FILES = ['Assignment2.java']
    JAVA_CLASSPATH = '.:./{}:/local/packages/jdbc-postgresql/postgresql-9.4.1212.jar'.format(PATH_TO_SOLUTION)
    NEED_ORACLE = {'homeownerRecommendation': False, 'booking': True}
    tester = JDBCTester(oracle_database=DATABASE, test_database=DATABASE, user_name=USER, user_password=PASSWORD,
                        data_files=DATA_FILES, schema_name=SCHEMA_NAME, path_to_solution=PATH_TO_SOLUTION,
                        java_files=JAVA_FILES, java_classpath=JAVA_CLASSPATH, need_oracle=NEED_ORACLE)
    tester.run()

File: A2/self_tests-JDBC/jdbc/JDBCTester.java
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;

public class JDBCTester {

	private String oracleDatabase;
	private String testDatabase;
	private String userName;
	private String userPassword;
	private String schemaName;

	private Assignment2Solution oracle;
	private Assignment2 test;

	private static void printResult(String name, String input, String expected, String actual, int marks, String status) {

		System.out.println(status + ": (java) " + name + " - " + actual);
	}

	private static void printResultFile(String testName, String message, String oracleResults, String testResults) {

		System.err.println("(java) " + testName + " - " + message);
		System.err.println(" Expected: " + oracleResults);
		System.err.println(" Actual: " + testResults + "\n");
	}

	public JDBCTester(String oracleDatabase, String testDatabase, String userName, String userPassword, String schemaName) {

		this.oracleDatabase = oracleDatabase;
		this.testDatabase = testDatabase;
		this.userName = userName;
		this.userPassword = userPassword;
		this.schemaName = schemaName;
	}

	private boolean initDB(String dataName) {

		try {
			this.oracle = new Assignment2Solution();
			this.test = new Assignment2();
			final String JDBC_PREAMBLE = "jdbc:postgresql://localhost:5432/";
			boolean testConnected = this.test.connectDB(JDBC_PREAMBLE + this.testDatabase, this.userName, this.userPassword);
			if (!testConnected || this.test.connection == null || !this.test.connection.isValid(0)) {
				JDBCTester.printResult("connectDB + " + dataName, "", "", "connectDB() failed", 0, "fail");
				return false;
			}
			JDBCTester.printResult("connectDB + " + dataName, "", "", "All good", 1, "pass");	
			return true;
		}
		catch (Exception e) {
			JDBCTester.printResult("connectDB + " + dataName, "", "", e.getMessage(), 0, "fail");
			return false;
		}
	}

	private void closeDB(String dataName) {

		try {
			boolean testDisconnected = this.test.disconnectDB();
			if (!testDisconnected || (this.test.connection != null && !this.test.connection.isClosed())) {
				JDBCTester.printResult("disconnectDB + " + dataName, "", "", "disconnectDB() failed", 0, "fail");
				return;
			}
			JDBCTester.printResult("disconnectDB + " + dataName, "", "", "All good", 1, "pass");
		}
		catch (Exception e) {
			JDBCTester.printResult("disconnectDB + " + dataName, "", "", e.getMessage(), 0, "fail");
		}
	}

	private void testHomeownerRecommendation(String testDataName) throws Exception {

		ArrayList testResult = this.test.homeownerRecommendation(4000);
		ArrayList oracleResult = this.oracle.homeownerRecommendation(4000);
		String message;
		int mark;
		String status;
		if (oracleResult.equals(testResult)) {
			message = "All good";
			mark = 1;
			status = "pass";
		}
		else {
			message = "homeownerRecommendation() output differs";
			mark = 0;
			status = "fail";
		}
		JDBCTester.printResult(testDataName, "", "", message, mark, status);
		JDBCTester.printResultFile(testDataName, message, Arrays.toString(oracleResult.toArray()), Arrays.toString(testResult.toArray()));
	}

	private void testBooking(String testDataName) throws Exception {

        SimpleDateFormat ft = new SimpleDateFormat("yyyy-MM-dd");
		boolean testBooked = this.test.booking(6000, ft.parse("2016-10-05"), 2, 120);
		boolean oracleBooked = this.oracle.booking(6000, ft.parse("2016-10-05"), 2, 120);
		String message;
		int mark;
		String status;
		if (oracleBooked == testBooked) {
			message = "All good";
			mark = 1;
			status = "pass";
		}
		else {
			message = "booking() output differs";
			mark = 0;
			status = "fail";
		}
		JDBCTester.printResult(testDataName, "", "", message, mark, status);
		JDBCTester.printResultFile(testDataName, message, String.valueOf(oracleBooked), String.valueOf(testBooked));
	}

	public void run(String dataFile, String testName) {

		String dataName = dataFile.split("\\.")[0];
		String testDataName = testName + " + " + dataName;
		if (!this.initDB(dataName)) {
			return;
		}
		try {
			switch (testName) {
				case "homeownerRecommendation":
					this.testHomeownerRecommendation(testDataName);
					break;
				case "booking":
					this.testBooking(testDataName);
					break;
			}
		}
		catch (Exception e) {
			JDBCTester.printResult(testDataName, "", "", e.getMessage(), 0, "error");
		}
		finally {
			this.closeDB(dataName);
		}
	}

	public static void main(String args[]) {

		final String DATABASE = args[0];
		final String USER = args[1];
		final String PASSWORD = args[2];
		final String SCHEMA_NAME = args[3];
		final String DATA_FILE = args[4];
		final String TEST_NAME = args[5];

		JDBCTester tester = new JDBCTester(DATABASE, DATABASE, USER, PASSWORD, SCHEMA_NAME);
		tester.run(DATA_FILE, TEST_NAME);
	}

}
